src/pudl/transform/classes.py

```diff
@@ -431,6 +431,60 @@ def correct_units(df: pd.DataFrame, params: UnitCorrections) -> pd.DataFrame:
 #####################################################################################
 # Abstract Table Transformer classes
 #####################################################################################
+def cache_df(key: str = "main") -> Callable[..., pd.DataFrame]:
+    """A decorator for caching dataframes within an :class:`AbstractTableTransformer`.
+
+    It's often useful during development or debugging to be able to track the evolution
+    of data as it passes through several transformation steps. Especially when some of
+    the steps are time consuming, it's nice to still get a copy of the last known state
+    of the data when a transform raises an exception and fails.
+
+    This decorator lets you easily save a copy of the dataframe being returned by a
+    class method for later reference, before moving on to the next step. Each unique key
+    used within a given :class:`AbstractTableTransformer` instance results in a new
+    dataframe being cached. Re-using the same key will overwrite previously cached
+    dataframes that were stored with that key.
+
+    Saving many intermediate steps can provide lots of detailed information, but will
+    use more memory. Updating the same cached dataframe as it successfully passes
+    through each step lets you access the last known state it had before an error
+    occurred.
+
+    This decorator requires that the decorated function return a single
+    :class:`pd.DataFrame`, but it can take any type of inputs.
+
+    Args:
+        key: The key that will be used to store and look up the cached dataframe in the
+            internal ``self._cached_dfs`` dictionary.
+
+    Returns:
+        The decorated class method.
+
+    """
+
+    def _decorator(func: Callable[..., pd.DataFrame]) -> Callable[..., pd.DataFrame]:
+        @wraps(func)
+        def _wrapper(self: AbstractTableTransformer, *args, **kwargs) -> pd.DataFrame:
+            df = func(self, *args, **kwargs)
```

Comment:
does a `df = func` because the `func` here is expected to return a dataframe? and the `*args` and `**kwargs` are passed in here and in `_wrapper` but not used in this wrapper because  if they weren't they wouldn't be passed into the actual function or method that this decorator will wrap?

Reply:
This is definitely confusing! And also apparently the way that decorators are typically written.

The decorator has to do its own work, based on its own arguments (here, caching a dataframe in a dictionary with a particular key), as well as ensuring that the function it is decorating does its work.

Decorators take a function as an argument and return a function. That returned function is then called by Python with the function that the decorator is applied to as its argument. Here the decorated method is `func`, which can take any inputs (as long as the first one is self, since it's a method) but has to return a single dataframe. The wrapper calls that method, passing in whatever arguments it got originally, and then does the decorator's work of caching the dataframe.

I'm probably not doing a good job of explaining this since I just barely understood it well enough to get it to work.

Reply:
I'll add a link to the decorator tutorial I learned from:

https://realpython.com/primer-on-python-decorators/#fancy-decorators